src/mapping/query/functions.rs

```diff
@@ -210,6 +211,161 @@ impl Function for ToTimestampFn {
+            _ => Err(r#"unable to apply "downcase" to non-string types"#.to_string()),
+        }
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(in crate::mapping) struct UuidV4Fn {}
+
+impl UuidV4Fn {
+    pub(in crate::mapping) fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Function for UuidV4Fn {
+    fn execute(&self, _: &Event) -> Result<Value> {
+        let mut buf = [0; 36];
+        let uuid = uuid::Uuid::new_v4().to_hyphenated().encode_lower(&mut buf);
+
+        Ok(Value::Bytes(Bytes::copy_from_slice(uuid.as_bytes())))
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Sha1Fn {
+    query: Box<dyn Function>,
+}
+
+impl Sha1Fn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for Sha1Fn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        use sha1::{Digest, Sha1};
+
+        match self.query.execute(ctx)? {
+            Value::Bytes(bytes) => {
+                let sha1 = hex::encode(Sha1::digest(&bytes));
+                Ok(Value::Bytes(sha1.into()))
+            }
+            _ => Err(r#"unable to apply "sha1" to non-string types"#.to_string()),
+        }
+    }
+}
+
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Md5Fn {
+    query: Box<dyn Function>,
+}
+
+impl Md5Fn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for Md5Fn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        use md5::{Digest, Md5};
+
+        match self.query.execute(ctx)? {
+            Value::Bytes(bytes) => {
+                let md5 = hex::encode(Md5::digest(&bytes));
+                Ok(Value::Bytes(md5.into()))
+            }
+            _ => Err(r#"unable to apply "md5" to non-string types"#.to_string()),
+        }
+    }
+}
+//------------------------------------------------------------------------------
+
+#[derive(Debug)]
+pub(in crate::mapping) struct NowFn {
+    query: Box<dyn Function>,
+}
+
+impl NowFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for NowFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        use chrono::{SecondsFormat, TimeZone, Utc};
+        use chrono_tz::Tz;
+
+        match self.query.execute(ctx)? {
+            Value::Bytes(bytes) => {
+                let utc = Utc::now().naive_utc();
+                let dt = std::str::from_utf8(&bytes)
+                    .map_err(|e| e.to_string())
+                    .and_then(str::parse::<Tz>)?
+                    .from_utc_datetime(&utc)
+                    .to_rfc3339_opts(SecondsFormat::Nanos, true);
+
+                Ok(Value::Bytes(dt.into()))
```

Comment:
Would it be more useful to return a Value::Timestamp here?

Reply:
I agree. We have types for timestamps, they should be used, particularly given the presence of a timestamp parse function that outputs `Value::Timestamp`.

Reply:
> Would it be more useful to return a Value::Timestamp here?

`Value::Timestamp` requires UTC-based timestamps, whereas this explicitly returns a timestamp with the provided time zone.

I'm not sure what the best solution to that is. I tried making the time zone generic:

```rust
enum Value<Tz: TimeZone = Utc> {
    Timestamp(DateTime<Tz>),
}
```

But that has a (huge?) ergonomics hit, as I now have to update a lot of function signatures to have Vector compile, which might not be worth the change?

Any other suggestions?

Note that I also have an open question if we actually want to accept a time zone parameter for this function (see: https://github.com/timberio/vector/issues/3753#issuecomment-692632132).

Reply:
^ The above comment is moot, as @binarylogic mentioned we're going to let `now` return `UTC` only for now: https://github.com/timberio/vector/issues/3753#issuecomment-692749276.